task_processing/plugins/mesos/resource_helpers.py

```diff
@@ -0,0 +1,69 @@
+from typing import Any
+from typing import Dict
+
+from task_processing.plugins.mesos.task_config import MesosTaskConfig
+
+NUMERIC_RESOURCES = frozenset(['cpus', 'mem', 'disk', 'gpus'])
+OTHER_RESOURCES: dict = {'ports': []}
+
+
+def _make_resource_dict():
+    d = dict([(r, 0) for r in NUMERIC_RESOURCES])
+    d.update(OTHER_RESOURCES)
+    return d
+
+
+def get_offer_resources(offer, role: str) -> Dict[str, Any]:
+    """ Get the resources from a Mesos offer
+
+    :param offer: the payload from a Mesos resourceOffer call
+    :param role: the Mesos role we want to get resources for
+    :returns: a mapping from resource name -> available resources for the offer
+    """
+    offer_resources = _make_resource_dict()
+    for resource in offer.resources:
+        if resource.role != role:
+            continue
+
+        if resource.name in NUMERIC_RESOURCES:
+            offer_resources[resource.name] += resource.scalar.value
+        elif resource.name == 'ports':
+            offer_resources['ports'] = [r for r in resource.ranges.range]
+
+    return offer_resources
+
+
+def allocate_task_resources(task: MesosTaskConfig, offer_resources: Dict[str, Any]) -> int:
```

Comment:
this function is a bit confusing: it mutates it's input and it returns a value that seems to be a single kind of "other" resource. i would prefer it to have a following form:

    consume_resources(
        requested_resources,
        offered_resources) -> (consumed_resources, remaining_resources)

where `consumed_resources` would mostly end up being a copy of `requested_resources` with special handling for `OTHER_RESOURCES` (like picking a port and whatnot) and `remaining_resources` would be used as an input in place of `offered_resources` for the next task. thoughts?

Reply:
sgtm!